### ISSUE TITLE
2.0.0-rc.3 / rc.4 bugs

### ISSUE THREAD

**Issue by cvn** [TTLabs/EvaporateJS#267]
I tested rc.3/rc.4 and ran into a couple bugs.

### r2.0.0-rc.4
- [x] When trying to initiate any upload I get this error. The error repeats periodically and the upload never starts.

```
initiate error: my-bucket/my-file.zip TypeError: con.cryptoHexEncodedHash256 is not a function(…)
```

I still wanted to try the new parallel upload feature, so I moved onto rc.3.

### r2.0.0-rc.3
- [ ] Resuming an upload `instance.resume(file_key)` doesn't seem to work. I get some output in the console, but the upload never actually resumes. I waited a few minutes.

```
my-app.js:141 PAUSE
evaporate.js:534 pausing FileUpload, force: true my-bucket/my-file.zip
evaporate.js:1061   ### 6291456
evaporate.js:1061   ### 6291456
evaporate.js:1061   ### 6291456
evaporate.js:1061   ### 6291456
evaporate.js:1061   ### 1527735
my-app.js:145 RESUME
evaporate.js:558 resuming FileUpload  my-bucket/my-file.zip
evaporate.js:1325 setupRequest() for uploadId: P2reJimZlBmthHrIKVUKVo92ifTK8_AgjVgyUqCU9CLQtESFGMacB1f2jBUs8mvjOTYDrWI8Z8e2gNu1zqrhnJcn1rrQEKfG0XXQA0YeN6VZV8ciVij6jHyOjiqN3KIw for part marker: 0
evaporate.js:1637 V2 stringToSign: GET
x-amz-date:Sun, 20 Nov 2016 03:26:32 GMT
/my-bucket/my-file.zip?uploadId=P2reJimZlBmthHrIKVUKVo92ifTK8_AgjVgyUqCU9CLQtESFGMacB1f2jBUs8mvjOTYDrWI8Z8e2gNu1zqrhnJcn1rrQEKfG0XXQA0YeN6VZV8ciVij6jHyOjiqN3KIw
evaporate.js:1186 get upload parts signature: 90PD2eX3x1giEWW8+l0Rh3A/7HI=
evaporate.js:1541 1 "payload bytes:" 0 "->" 6291456 ", size:" 6291456
evaporate.js:1541 2 "payload bytes:" 6291456 "->" 12582912 ", size:" 6291456
evaporate.js:1541 3 "payload bytes:" 12582912 "->" 18874368 ", size:" 6291456
evaporate.js:1541 4 "payload bytes:" 18874368 "->" 25165824 ", size:" 6291456
evaporate.js:1541 5 "payload bytes:" 25165824 "->" 31457280 ", size:" 1527735
```

**Comment by bikeath1337** [comment]
Thank you for the feedback, @cvn ! I will  investigate and provide fixes, where necessary.

**Comment by bikeath1337** [comment]
Regarding the reported issue with rc.4. The Changelog stated that the awsSignatureVersion default changes from '2' to '4'. So it appears you are using version 2. To work with rc.4, just configure Evaporate to specifically set the version like so and give it a try:

```
{
awsSignatureVersion: '2'
}
```

**Comment by bikeath1337** [comment]
Thanks again for the feedback. In my rc.5 release candidate (not published), I found issues with pause resume that I will address.

Curious to know if you were force pausing or just pausing?

**Comment by cvn** [comment]
Force pausing.

Also, I tried updating to v4 signing, to see if it would make any difference, but I got the same results.

**Comment by bikeath1337** [comment]
Thanks for confirming about Force pausing. The problem is related to the subtle difference between Force pausing and Canceling an upload. Will address the issue.

**Comment by bikeath1337** [comment]
Just released rc5 with the fixes to Pause/Resume and support for pluggable signing methods.

**Comment by cvn** [comment]
Night and Day. I tested rc.5 for several hours, and it's been very reliable in the face of pause/resume abuse. No errors at all, and only a few isolated issues with resuming.

Here are 2 scenarios that where resume does not work in rc.5:

1. Start two 100MB uploads
    - upload1 starts uploading, upload2 is queued
2. pause all `instance.pause(undefined, {force: true})`
3. resume all `instance.resume()`
    - upload1 starts uploading again
4. pause upload1 `instance.pause(upload1_key, {force: true})`
    - upload2 does not start uploading

Scenario 2

1. Start two 100MB uploads
    - upload1 starts uploading, upload2 is queued
2. pause upload1 `instance.pause(upload1_key, {force: true})`
    - upload2 starts uploading
3. resume upload1 `instance.resume(upload1_key)`
4. pause upload2 `instance.pause(upload2_key, {force: true})`
    - upload1 starts uploading
5. resume upload2 `instance.resume(upload2_key)`
6. pause upload1 `instance.pause(upload1_key, {force: true})`
    - upload2 does not start uploading

In Scenario 2, pausing and resuming upload2 will cause it to start uploading again, but with Scenario 1 I've found no way to restart upload2.

**Comment by bikeath1337** [comment]
@cvn thanks for the clear instructions. They should both be resolved in rc.6, which I just published.

**Comment by cvn** [comment]
I just tested rc.6 for a few hours. The pause/resume behavior is flawless now. But parallel uploads seem to not be working. Files are uploading sequentially for me, like pre-2.0.

**Comment by cvn** [comment]
After more testing, parallel uploads do work in rc.6, but something's a little off.

Now, when you upload multiple files, only one file will start uploading, and the the next file does not start until there is an ETag event in the console (a part completing, I think). This affects uploading sequences of files differently depending on their size.

- `size < partSize`
    - uploads are sequential
    - were parallel up to maxParts in rc.5
- `partSize < size < maxParts * partSize`
    - uploads start sequentially, but become parallel as ETag events happen
    - started parallel in rc.5
- `size > maxParts * partSize`
    - behaves similar to rc.5

**Comment by bikeath1337** [comment]
@cvn I don't have time to publish the rc candidate with the fix tonight but if you want, you can take evaporate.js from https://github.com/bikeath1337/EvaporateJS/tree/promises.

**Comment by cvn** [comment]
Cool. rc.7 from your branch fixes the issue with parallel uploads for small files. I found one very specific regression from rc.6:

1. Start 100MB upload1
2. Force pause upload1
3. Start 100MB upload2
4. Resume upload1
    - upload1 does not resume when upload2 parts complete

upload1 can be resumed by pausing and resuming it once there are available slots, but this was unnecessary in rc.6.

**Comment by bikeath1337** [comment]
Thanks much. Interestingly, these issues are all related to code smells that I had identified but didn't know how to properly address. If I knew of a way to build a test for this, I would have, but my skills with Ava/Sinon aren't good enough yet. Will address and report back.

**Comment by bikeath1337** [comment]
@cvn rc7 is available.

**Comment by cvn** [comment]
With the new rc.7 I have run into an issue where after cancelling uploads, uploading breaks.

1. Start two 100MB uploads
   - upload1 starts uploading, upload2 is queued
2. cancel upload1, then upload2 rapidly
    - In my app, I call cancel after deleting stubs in parallel on the server, <100ms between events is usual
    - This logs twice: `Requeued my-bucket/upload2.jpg status: 20 reason: file done`
3. Start upload3
    - This logs: `Requeued my-bucket/upload2.jpg status: 20 reason: first file`
    - upload3 does not start

**Comment by cvn** [comment]
This is not a bug, but it's related to rc.7.

I maintain my own upload queue outside of evaporate. I only want to add files to evaporate if there are available upload slots. In the previous rc's, I was using `if (instance.getRemainingSlots(true)) { add next file }` in the first progress event of an upload. This worked well, and with large uploads, it wouldn't add the next file yet, but with small uploads, it would keep adding files until the slots were full.

In rc.7, getRemainingSlots moved, so now I'm using `if (instance.evaporatingCount < instance.config.maxConcurrentParts) {...}`. But when I run this in the first progress event of an upload, the evaporatingCount is 1, even with 100MB files. So it always adds the next file.

Is there a better way to do this? If not, would you consider supporting this use case?

**Comment by bikeath1337** [comment]
Though I sympathize with your issue, if you're using Evaporate internals, this isn't something I can help you with per se as we need to be able to change its implementation without impacting users.

However, if there is a need that we can meet and it's one that Evaporate should manage, I'm all ears but it sounds like you're not letting Evaporate do it's work as it's the only one in control of the available slots. Why can't you let Evaporate manage the queue? That's its designed job.

Why don't you open a new issue and we can go from there. Thanks!

**Comment by bikeath1337** [comment]
Closing this as the release of v2.0 is nigh. :)

**Comment by cvn** [comment]
The latest issue I mentioned [above](https://github.com/TTLabs/EvaporateJS/issues/267#issuecomment-263769325), about problems with cancel, is still unresolved. It might be worth keeping this open till that's resolved. I can make a new ticket for it, if that helps.

**Comment by bikeath1337** [comment]
That would be great. Thank you.
